### PR TITLE
[WIP TTS] | Aborted Transactions Cache Hack

### PR DESCRIPTION
**Pull request by sudiksha27**
## General
Adds a new interfaces - `KnownCommittedTimestamps` which in turn relies on `KnownAbortedTimestamps`. 

**After this PR**:
Introduces new interface - `KnownCommittedTimestamps` which in turn relies on `KnownAbortedTimestamps`. 
`KnownAbortedTimestamps` uses bucketed caching. 
KnownCommittedTimestamps is not complete - broadly the algorithm is to look in knownConcluded transactions and fall back on transactions table as the source of truth.

**Priority**:
Before end of week

**Concerns / possible downsides (what feedback would you like?)**:
Is there a race condition I missed in `abortedTimestamps` caching?
Is the case where transaction has not been concluded handled correctly?

**Is documentation needed?**:
Docs added

## Compatibility
~**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:~

~**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:~

~**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:~

~**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:`~

~**Does this PR need a schema migration?**~

## Testing and Correctness
~**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:~

**What was existing testing like? What have you done to improve it?**:
In progress

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:
For `KnownAbortedTransactions`, we have bucketed range of timestamps with bucket size = 1million
A bucket is cached upon access. 
A bucket is considered immutable and hence reliable iff greatest timestamp in that bucket has been concluded.
For mutable buckets, it is possible that the set of aborted txns can change which is why we refresh the cache.

We can add another optimization -> check if the highest known aborted > timestamp; if so, we do not have to refresh cache.

~**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:~

## Execution
~**How would I tell this PR works in production? (Metrics, logs, etc.)**:~

~**Has the safety of all log arguments been decided correctly?**:~

**Will this change significantly affect our spending on metrics or logs?**:
No

**How would I tell that this PR does not work in production? (monitors, etc.)**:
This PR does not do any wiring and is therefore not actually functioning.

~**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:~

~**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:~

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:
The write of `abortedTxns` component should not be called with high concurrency.
The reads on `abortedTxns` and `committedTxns` can be highly concurrent which is why we have introduced caching for aborted transactions

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:
I do not expect database calls to be expensive.
It might be possible that the knownConcluded transactions are not up to date and we need to make db calls to txn table in addition to concludedTxnStore
Another point of contention is that it might take a while for bucket of aborted txns to be concluded which would mean a remote read  for every transaction that has not aborted in the worst case. 

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:
I do not _think_ so 

## Development Process
**Where should we start reviewing?**:
KnownAbortedTransactions

~**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:~

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownAbortedTransactions.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Weigher;
+import com.palantir.atlasdb.AtlasDbConstants;
+import com.palantir.atlasdb.transaction.knowledge.KnownConcludedTransactions.Consistency;
+import java.util.Comparator;
+import java.util.Set;
+import org.checkerframework.checker.index.qual.NonNegative;
+
+public class DefaultKnownAbortedTransactions implements KnownAbortedTransactions {
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownAbortedTransactions.java:28]: Atlas convention is to call it `KnownAbortedTransactionsImpl`, but we haven't been super consistent so this is fine

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/KnownAbortedTransactions.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.knowledge;
+
+import java.util.Set;
+
+/**
+ * Represents a set of start timestamps that belong to transactions that are known to have aborted.
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/KnownAbortedTransactions.java:22]: *have been

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/KnownAbortedTransactions.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ */
+public interface KnownAbortedTransactions {
+
+    /**
+     * Returns whether the transaction that started at the provided timestamp is known to have aborted.
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/KnownAbortedTransactions.java:27]: Maybe reword a bit: For a concluded transaction with the given start timestamp, returns true if it is known that the transaction has been aborted. Calling this method for a transaction that has not been concluded is undefined

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/KnownAbortedTransactions.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+     * This method should only be called for transactions that are known to have concluded.
+     *
+     * @param startTimestamp start timestamp associated with the value we are checking for
+     * @return whether the transaction that started at the provided timestamp is known to have aborted.
+     */
+    boolean isKnownAborted(long startTimestamp);
+
+    /**
+     * Registers the fact that any transactions that had startTimestamp in set of timestamps provided have been
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/KnownAbortedTransactions.java:36]: Probably should not be exposed if it's not meant to be called

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownCommittedTransactions.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        this.transactionService = transactionService;
+    }
+
+    @Override
+    public boolean isKnownCommitted(long startTimestamp) {
+        if (!isConcluded(startTimestamp)) {
+            return getTransactionStateFromTransactionsTable(startTimestamp);
+        }
+
+        return !knownAbortedTransactions.isKnownAborted(startTimestamp);
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownCommittedTransactions.java:43]: We can't treat aborted the same as in progress -- we might want to abort in progress explicitly, but we shouldn't do the same for already aborted.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownCommittedTransactions.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            TransactionService transactionService) {
+        this.knownConcludedTransactions = knownConcludedTransactions;
+        this.knownAbortedTransactions = knownAbortedTransactions;
+        this.transactionService = transactionService;
+    }
+
+    @Override
+    public boolean isKnownCommitted(long startTimestamp) {
+        if (!isConcluded(startTimestamp)) {
+            return getTransactionStateFromTransactionsTable(startTimestamp);
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownCommittedTransactions.java:40]: We want to:
1) if we already know, return true (also for aborted)
2) if we don't know, atomically refresh and get
3) if now it is deleted, throw

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownAbortedTransactions.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    private static Cache<Long, Set<Long>> cache() {
+        return Caffeine.newBuilder()
+                .maximumWeight(100_000)
+                .weigher(new EntryWeigher())
+                .build();
+    }
+
+    private static final class EntryWeigher implements Weigher<Long, Set<Long>> {
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownAbortedTransactions.java:100]: nit: descriptive name

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownAbortedTransactions.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        if (canBeReliablyCached) {
+            softCache.invalidate(bucket);
+            reliableCache.get(bucket, this::getAbortedTransactionsRemote);
+        }
+        return softCache.get(bucket, this::getAbortedTransactionsRemote);
+    }
+
+    private boolean bucketCanBeReliablyCached(long bucket) {
+        long maxTsInCurrentBucket = ((bucket + 1) * AtlasDbConstants.ABORTED_TIMESTAMPS_BUCKET_SIZE) - 1;
+        return !knownConcludedTransactions.isKnownConcluded(maxTsInCurrentBucket, Consistency.REMOTE_READ);
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownAbortedTransactions.java:82]: why is this negated?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownAbortedTransactions.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+import java.util.Set;
+import org.checkerframework.checker.index.qual.NonNegative;
+
+public class DefaultKnownAbortedTransactions implements KnownAbortedTransactions {
+    private final KnownConcludedTransactions knownConcludedTransactions;
+    private final AbortedTimestampStore abortedTimestampStore;
+    private final Cache<Long, Set<Long>> reliableCache;
+    private final Cache<Long, Set<Long>> softCache;
+
+    public DefaultKnownAbortedTransactions(
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownAbortedTransactions.java:34]: Some thoughts:

There should always be at most one bucket that cannot be reliably cached. Let's keep these in a separate `ConcurrentHashSet` or something.
When we get a query:
- if it's in one of the cached buckets, we just return
- if it's in the funny bucket, since `isKnownAborted` must be only be called when it is known to be concluded, we can compare the queried timestamp to the greatest known committed since the last update for this bucket. If it is greater, we can update the state of this one set with the diff since last update and then return
- if it's in a greater bucket, then we likely want to update the funny bucket to the end, push it into the cache, reset the funny set, and then decide for this ts what to do

**Review by gmaretic — COMMENTED** [review]
I'm not fully following the logic in `DKAT`, happy to discuss offline

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownCommittedTransactions.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        return !knownAbortedTransactions.isKnownAborted(startTimestamp);
+    }
+
+    private boolean getTransactionStateFromTransactionsTable(long startTimestamp) {
+        /**
+         * We have not wired in the transactionService#safeGet() api. Broad idea will be:
+         * case committed ->  return true
+         * case inProgress -> return false
+         * case deleted -> refresh knownConcluded and throw
+         * */
```
Comment (sudiksha27) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownCommittedTransactions.java:52]: do not throw on deleted

Comment (sudiksha27) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownCommittedTransactions.java:52]: 1. local
2. txn table; if deleted refresh knownConcluded (can also have caching for this one)
3. remote if 2 did not return

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownAbortedTransactions.java (hunk 8 above)
Comment (sudiksha27) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownAbortedTransactions.java:82]: my absolute bad

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownCommittedTransactions.java (hunk 6 above)
Comment (sudiksha27) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownCommittedTransactions.java:40]: paired online: this is slightly different to what we ended up deciding
